lib/oxide-vpc/tests/integration_tests.rs

```diff
@@ -981,37 +1052,122 @@ fn unpack_and_verify_icmp4(
     assert_eq!(icmp.echo_seq_no(), seq_no);
 }
 
+fn unpack_and_verify_icmp6(
+    pkt: &Packet<Parsed>,
+    expected_ident: u16,
+    seq_no: u16,
+    encapped: bool,
+    body_seg: usize,
+    src_ip: Ipv6Addr,
+    dst_ip: Ipv6Addr,
+) {
+    // Length is factored into pseudo header calc.
+    // We know there are no ext headers.
+    let pay_len = pkt.meta().inner_ip6().unwrap().pay_len as usize;
+
+    let src_ip = smoltcp::wire::Ipv6Address::from(src_ip).into();
+    let dst_ip = smoltcp::wire::Ipv6Address::from(dst_ip).into();
+
+    let icmp_offset = pkt.body_offset() - IcmpHdr::SIZE;
+    let tgt_offset = IP6_SZ + if encapped { VPC_ENCAP_SZ } else { 0 };
+    assert_eq!(icmp_offset, tgt_offset);
+    assert_eq!(pkt.body_seg(), body_seg);
+
+    // Because we treat ICMPv6 as a full-fledged ULP, we need to
+    // unsplit the emitted header from the body.
+    let pkt_bytes = pkt.all_bytes();
+    let icmp = Icmpv6Packet::new_checked(&pkt_bytes[icmp_offset..][..pay_len])
+        .unwrap();
+    let parsy = Icmpv6Repr::parse(
+        &src_ip,
+        &dst_ip,
+        &icmp,
+        &smoltcp::phy::ChecksumCapabilities::ignored(),
+    )
+    .unwrap();
+
+    let mut emit_caps = smoltcp::phy::ChecksumCapabilities::ignored();
+    emit_caps.icmpv6 = smoltcp::phy::Checksum::Both;
+    let mut new_buf = vec![0; pay_len];
+    let mut icmp2 = Icmpv6Packet::new_unchecked(&mut new_buf[..]);
+    parsy.emit(&src_ip, &dst_ip, &mut icmp2, &emit_caps);
+
+    assert!(icmp.verify_checksum(&src_ip, &dst_ip));
+    assert_eq!(icmp.echo_ident(), expected_ident);
+    assert_eq!(icmp.echo_seq_no(), seq_no);
+}
+
 // Verify that an ICMP Echo request has its identifier rewritten by
 // SNAT.
 #[test]
 fn snat_icmp4_echo_rewrite() {
+    let dst_ip: Ipv4Addr = "45.55.45.205".parse().unwrap();
+    snat_icmp_shared_echo_rewrite(dst_ip.into());
```

Comment:
Don't need to detour through `Ipv4Addr`
```suggestion
    snat_icmp_shared_echo_rewrite("45.55.45.205".parse().unwrap());
```